### DISCUSSION TITLE
What would you ask the bot if the bot can read your project?

### DISCUSSION THREAD

**Discussion by intitni** [intitni/CopilotForXcode#366] [Ideas]
Starting from [0.24.0](https://github.com/intitni/CopilotForXcode/releases/tag/0.24.0), I will be working on providing the bot with project knowledge, starting with chat. 

**It would be great if you could share with me the questions you want the bot to answer about a project so that I know how to test the feature.**

The current implementation (in 0.24.0) is just a proof of concept that doesn't work well. It is done by function calling and ReAct Agent, which seems unsuitable for this task. I will try using RAG in future releases.

The project knowledge will (if doable) also come to the other features later.

**Comment by ICULikeMac** [comment]
Cool idea. What are your thoughts when the project is quite large? Will this be an issue? Or is this more for smaller projects? 

I'll have a think of some questions but it would probably be stuff between swift & objc which I'm not as familiar with.

**Reply by intitni** [reply]
It will use various retrieval methods to find the relevant pieces of information to look at, so it would be fine even if the project is large.

(But the current implementation may do something wrong if it's trying to read markdown files and you have a lot of markdown files in your project (´･ω･｀))

**Reply by ICULikeMac** [reply]
Fantastic, great work. 

I'll start noting down some questions I would ask it and post them here.

Thanks as always.

**Comment by gsbernstein** [comment]
This would be particularly helpful if you could ask it to make changes directly in the code, for example saying "can you consolidate usages of these networking implementations" or "can you change this feature to pass around UUIDs instead of strings?"

Even if it can't do that, maybe other useful things would be like "can you find where an anonymous closure could be retaining this class" or "is there anything that uses these URL constants that isn't reset on an environment change?"

**Reply by intitni** [reply]
The first two will be another feature that I would love to try in the future. 

For "can you find where an anonymous closure could be retaining this class", I think it's possible if the class is explicitly used in the closure. But I am not sure.

What does "reset on an environment change" mean?

**Reply by gsbernstein** [reply]
Was just using examples of things I'd been doing recently. For our app, we let devs switch between prod, staging, and other development servers from inside the app. When you do so, we log the user out and reset all the singletons that hold a reference to the base URL and other constants. I'm kinda glad I looked at it myself since it gave me the opportunity to centralize some of those references, but it's the kind of thing where an AI might be able to help track down downstream effects or make sure were refreshing them in the right order.

**Comment by Vybo** [comment]
Would it be feasible to add something like `@package` scope that would use only the swift package of the currently opened file as a scope? 

Using the whole project wouldn't fit even the 32k model for bigger apps, however some architectures use heavy SPM splitting, so that would help out. Also, most packages have a clear public interface for interacting with them.

**Reply by intitni** [reply]
I think it's fine given how RAG works. It won't put the whole code base into context, instead, it will try to find the most relevant pieces of code, which is the hard part. 

But yes it's a good idea that we should give the contents from the same package a higher priority so they are less likely to be cut off when we hit the context limit.

**Comment by intitni** [comment]
0.25.0 has been released with RAG-based project scope. However, there is still some additional work to be done.


https://github.com/intitni/CopilotForXcode/assets/793147/928ac8fc-18c1-4cf6-8b7b-149dace536a7


It seems to be usable even with models with 4k tokens. Currently, the retrieved content can only utilize half of the maximum tokens, allowing the model to generate a complete reply.

It works well only when the symbol name is correctly spelled (or partly spelled) in your message or in the chat history. It will not search for `Config` if you mention `Settings`. 

The agent thing has been removed, but I am considering bringing it back in the future to enable the bot to perform more sophisticated searches within the project.

Would it be a good idea to allow using a different model when project scope is on?
